src/server/game/Spells/Spell.cpp

```diff
@@ -1283,6 +1283,207 @@ void Spell::SelectImplicitCasterDestTargets(SpellEffIndex effIndex, SpellImplici
             dest = SpellDestination(x, y, liquidLevel, m_caster->GetOrientation());
             break;
         }
+        case TARGET_DEST_CASTER_FRONT_LEAP:
+        {
+            float distance = m_spellInfo->Effects[effIndex].CalcRadius();
+            Map* map = m_caster->GetMap();
+            uint32 phasemask = m_caster->GetPhaseMask();
+            float x, y, z;
+            float destx, desty, destz, ground;
+            float orientation = m_caster->GetOrientation();
+
+            m_caster->GetPosition(x, y, z);
+            destx = x + distance * cos(orientation);
+            desty = y + distance * sin(orientation);
+
+            ground = map->GetHeight(phasemask, destx, desty, z + 5.0f);
+            destz = ground + 0.5f;
+
+            if (!m_caster->HasUnitMovementFlag(MOVEMENTFLAG_FALLING) || (z - ground < 25.0f))
+            {
+                if ((m_caster->HasUnitMovementFlag(MOVEMENTFLAG_FALLING) && (z - ground > 3.0f)) && (!map->IsInWater(x, y, z)))
+                {
+                    Position pos;
+                    pos.Relocate(destx, desty, destz, orientation);
+                    m_caster->GetFirstCollisionPosition(m_caster->GetDistance(pos.GetPositionX(), pos.GetPositionY(), pos.GetPositionZ()), 0.0f);
+                    dest = SpellDestination(pos.GetPositionX(), pos.GetPositionY(), pos.GetPositionZ(), pos.GetOrientation());
+                    //float range = sqrt((desty - y)*(desty - y) + (destx - x)*(destx - x));
+                    //TC_LOG_ERROR("server", "Blink number 1, from falling, distance of blink = %f", range);
+                }
+                else
+                {
+                    /*//recalculate, we wanna fly above abyss and don't wanna stop to edge =) (if nessesary enable "stop to edge" - comment this lines
```

Comment:
Remove all the commented out code if it's not necessary. This code is simply making things more confusing